internal/servers/controller/listeners.go

```diff
@@ -161,61 +159,108 @@ func (c *Controller) configureForCluster(ln *base.ServerListener) (func(), error
 	return func() { go ln.GrpcServer.Serve(ln.ALPNListener) }, nil
 }
 
-func (c *Controller) stopListeners(serversOnly bool) error {
-	serverWg := new(sync.WaitGroup)
-	for _, ln := range c.conf.Listeners {
-		localLn := ln
-		serverWg.Add(1)
-		go func() {
-			defer serverWg.Done()
-
-			shutdownKill, shutdownKillCancel := context.WithTimeout(c.baseContext, localLn.Config.MaxRequestDuration)
-			defer shutdownKillCancel()
-
-			if localLn.GrpcServer != nil {
-				// Deal with the worst case
-				go func() {
-					<-shutdownKill.Done()
-					localLn.GrpcServer.Stop()
-				}()
-				localLn.GrpcServer.GracefulStop()
-			}
-			if localLn.HTTPServer != nil {
-				localLn.HTTPServer.Shutdown(shutdownKill)
-			}
-		}()
+func (c *Controller) stopServersAndListeners() error {
+	var mg multierror.Group
+	mg.Go(c.stopClusterGrpcServerAndListener)
+	mg.Go(c.stopHttpServersAndListeners)
+	mg.Go(c.stopApiGrpcServerAndListener)
+
+	stopErrors := mg.Wait()
+
+	err := c.stopAnyListeners()
+	if err != nil {
+		stopErrors = multierror.Append(stopErrors, err)
+	}
+
+	return stopErrors.ErrorOrNil()
+}
+
+func (c *Controller) stopClusterGrpcServerAndListener() error {
+	if c.clusterListener == nil {
+		return nil
 	}
+	if c.clusterListener.GrpcServer == nil {
+		return fmt.Errorf("no cluster grpc server")
+	}
+	if c.clusterListener.Mux == nil {
+		return fmt.Errorf("no cluster listener mux")
+	}
+
+	c.clusterListener.GrpcServer.GracefulStop()
+	err := c.clusterListener.Mux.Close()
+	return listenerCloseErrorCheck(c.clusterListener.Config.Type, err)
+}
+
+func (c *Controller) stopHttpServersAndListeners() error {
+	var closeErrors *multierror.Error
+	for i := range c.apiListeners {
+		ln := c.apiListeners[i]
```

Comment:
Is there a good reason to use indexing here instead of using a loop variable?

Reply:
I think using a loop variable here would work just fine.

I think I got confused because of [this](https://github.com/hashicorp/boundary/blob/main/internal/servers/controller/listeners.go#L174-L175) and not being entirely sure as to why it was done like that. 

The only reason I can think of is because we're using `localLn` in goroutines that spawn as part of a loop, there might be some pointer problems there where using `ln` would change under all goroutines and they'd end up only working on the last element of the slice? I'm not sure this would even be the case.

I know there's some [pitfalls](https://medium.com/codezillas/uh-ohs-in-go-slice-of-pointers-c0a30669feee) when working with slices of pointer types in go though.